pkg/distributors/contracts/MultiRewards.sol

```diff
@@ -70,191 +75,294 @@ contract MultiRewards is IMultiRewards, IDistributor, ReentrancyGuard, Temporari
+    ) public view returns (uint256) {
+        return Math.min(block.timestamp, rewardData[pool][rewarder][rewardsToken].periodFinish);
     }
 
-    function rewardPerToken(IERC20 stakingToken, IERC20 rewardsToken) public view returns (uint256) {
-        if (_totalSupply[stakingToken] == 0) {
-            return rewardData[stakingToken][rewardsToken].rewardPerTokenStored;
+    /**
+     * @notice Calculates the amount of reward per staked bpt that is
+     */
+    function rewardPerToken(
+        IERC20 pool,
+        address rewarder,
+        IERC20 rewardsToken
+    ) public view returns (uint256) {
+        if (_totalSupply[pool] == 0) {
+            return rewardData[pool][rewarder][rewardsToken].rewardPerTokenStored;
         }
         return
-            rewardData[stakingToken][rewardsToken].rewardPerTokenStored.add(
-                lastTimeRewardApplicable(stakingToken, rewardsToken)
-                    .sub(rewardData[stakingToken][rewardsToken].lastUpdateTime)
-                    .mulDown(rewardData[stakingToken][rewardsToken].rewardRate)
-                    .mulDown(1e18)
-                    .divDown(_totalSupply[stakingToken])
+            rewardData[pool][rewarder][rewardsToken].rewardPerTokenStored.add(
+                lastTimeRewardApplicable(pool, rewarder, rewardsToken)
+                    .sub(rewardData[pool][rewarder][rewardsToken].lastUpdateTime)
+                    .mulDown(rewardData[pool][rewarder][rewardsToken].rewardRate)
+                    .divDown(_totalSupply[pool])
             );
     }
 
     /**
      * @notice Calculates the amount of `rewardsToken` that `account` is able to claim
+     * from a particular rewarder
      */
     function earned(
-        IERC20 stakingToken,
+        IERC20 pool,
+        address rewarder,
         address account,
         IERC20 rewardsToken
     ) public view returns (uint256) {
         return
-            _balances[stakingToken][account]
+            _balances[pool][account]
                 .mulDown(
-                rewardPerToken(stakingToken, rewardsToken).sub(
-                    userRewardPerTokenPaid[stakingToken][account][rewardsToken]
+                rewardPerToken(pool, rewarder, rewardsToken).sub(
+                    userRewardPerTokenPaid[pool][rewarder][account][rewardsToken]
                 )
             )
-                .divDown(1e18)
-                .add(rewards[stakingToken][account][rewardsToken]);
+                .add(rewards[pool][rewarder][account][rewardsToken]);
+    }
+
+    /**
+     * @notice Calculates the total amount of `rewardsToken` that `account` is able to claim
+     */
+    function totalEarned(
+        IERC20 pool,
+        address account,
+        IERC20 rewardsToken
+    ) public view returns (uint256 total) {
+        for (uint256 r; r < _rewarders[pool][rewardsToken].length(); r++) {
+            total = total.add(earned(pool, _rewarders[pool][rewardsToken].at(r), account, rewardsToken));
```

Comment:
```suggestion
        uint256 length = _rewarders[pool][rewardsToken].length();
        for (uint256 r; r < length; r++) {
            total = total.add(earned(pool, _rewarders[pool][rewardsToken].unchecked_at(r), account, rewardsToken));
```

This will perform much fewer storage accesses.

Reply:
Doing 

```
EnumerableSet storage set = _rewarders[pool][rewardsToken];
for (...) {
  total = foo(set.at(r));
}
```

likely also saves gas - I'd be surprised if solc saved the result of the mapping computation.